### ISSUE TITLE
Determine best way to load client-side scripts

### ISSUE THREAD

**Issue by iainbryson** [dtu-compute/dtu-quiz#216]
Right now, the Web site and the quiz use two distinct methods of loading clientside scripts: the quiz uses gems and the Web site uses bower and serves up the scripts statically.

Neither site uses CDNs, which I suspect is the best option.  Possibly with a fallback to the local scripts if they fail (e.g. fallback.js).

We should perform an analysis to figure out what the fastest way of doing this is.  My current assumptions are:

(1) CDNs are fastest, would relieve pressure from our servers and would be best for caching
(2) Serving static files (like the Web site does) is second best.  NGINX is better at serving static files than the Ruby thin or the rails infrastructure. And these files can also be gzipped for faster transport
(3) Serving static clientside files from rails gems is the slowest (e.g. the Mathjax gem in used in the quiz). 

The other thing going on is the asset pipeline; what (if anything) should be concatenated and minified together.  Obviously this pushes in the opposite direction of the CDN, which keeps all the libraries separated.

In summary: we ought to test the performance of each of these methods to determine the best overall organization for speed (using Google PageSpeed and Yahoo YSlow)

**Comment by dtu-compute** [comment]
Very very interesting read!

I suppose delivering the quizzes is the biggest bottle neck? The automatic forced handing in should be a lot less intense as as student would have handed in before the time runs out.

So for our needs if it can serve 1500 quizzes in 60 sec., then it would be acceptable. If it e.g. can can deliver 1500 quizzes in 10 sec. then we should perhaps burn some performance on making the code cleaner and easier to maintain if it have some complicated performace optimisations.

At some point we talked about letting the students get the quiz in a encrypted state, so they can get it before the quiz starts and hence reduce the server load, and when the clock hits begin time the decryption key is sent to each client. Would this make the code very complicated?

It would be super interesting to know how close/far we are from 1500 quizzes / 60 sec. to determine what is needed. Are you familiar with commercial services that we can use to perform such tests? I imagine they need a hook into the quiz to login, take the quiz, click randomly, and submit.?

**Comment by dtu-compute** [comment]
> I do notice in the logs a good amount of traffic about MathJax etc… which seem like a pure waste for the rails app to deal with. It’s inefficient and makes debugging messier.

What would the preferred be to do instead?

**Comment by iainbryson** [comment]
The way the course Web site does it is via bower/gulp frontend pipeline.  All the frontend scripts — including MathJax — are compiled (or loaded off CDN).  All the frontend files are then placed on a shared volume for NGINX to serve as static files.

This leaves Rails handling only the dynamic content, which is what it is good at.

**Comment by dtu-compute** [comment]
Ok, then we will go that that as well for the quiz.
